prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt

```diff
@@ -0,0 +1,192 @@
+package gov.cdc.prime.router.fhirengine.translation.hl7
+
+import ca.uhn.hl7v2.HL7Exception
+import ca.uhn.hl7v2.model.Message
+import ca.uhn.hl7v2.util.Terser
+import gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchema
+import gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchemaElement
+import gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchemaReader
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathUtils
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.HL7Utils
+import org.apache.logging.log4j.kotlin.Logging
+import org.hl7.fhir.r4.model.Base
+import org.hl7.fhir.r4.model.Bundle
+
+/**
+ * Convert a FHIR [bundle] to an HL7 message using the [schemaRef] to perform the conversion.
+ * The converter will error out if [strict] is set to true and there is an error during the conversion.  if [strict]
+ * is set to false (the default) then any conversion errors are logged as a warning.  Note [strict] does not affect
+ * the schema validation process.
+ * @property terser the terser to use for building the HL7 message (use for dependency injection)
+ */
+class FhirToHl7Converter(
+    private val bundle: Bundle,
+    private val schemaRef: ConfigSchema,
+    private val strict: Boolean = false,
+    private var terser: Terser? = null
+) : Logging {
+    /**
+     * Convert a FHIR [bundle] to an HL7 message using the [schema] in the [schemaFolder] location to perform the conversion.
+     * The converter will error out if [strict] is set to true and there is an error during the conversion.  if [strict]
+     * is set to false (the default) then any conversion errors are logged as a warning.  Note [strict] does not affect
+     * the schema validation process.
+     * @property terser the terser to use for building the HL7 message (use for dependency injection)
+     */
+    constructor(
+        bundle: Bundle,
+        schema: String,
+        schemaFolder: String,
+        strict: Boolean = false,
+        terser: Terser? = null
+    ) : this(bundle, ConfigSchemaReader.fromFile(schema, schemaFolder), strict, terser)
+
+    /**
+     * Convert the given bundle to an HL7 message.
+     * @return the HL7 message
+     */
+    fun convert(): Message {
+        // Sanity check, but the schema is assumed good to go here
+        check(!schemaRef.hl7Type.isNullOrBlank())
+        check(!schemaRef.hl7Version.isNullOrBlank())
+        val message = HL7Utils.SupportedMessages.getMessageInstance(schemaRef.hl7Type!!, schemaRef.hl7Version!!)
+
+        // Sanity check, but at this point we know we have a good schema
+        check(message != null)
+        terser = Terser(message)
+        processSchema(schemaRef, bundle)
```

Comment:
Why pass schemaRef and bundle here since they are class variables anyway and the processSchema function is an internal class function?

Reply:
Because that function processSchema is used recursively in the processElement function

Reply:
Bundle because in the recursive call we can pass the focus resource

Reply:
Yep, that makes sense.